### ISSUE TITLE
golangci-lint does not play nice in a check derviation

### ISSUE THREAD

**Issue by colin-arnott-xero** [NixOS/nixpkgs#202614]
### Describe the bug
while trying to write a flake check for `golangci-lint`, it became apparent that, while the binary is built correctly, the core structure of `golangci-lint` and `go` break fundamental assumptions about sandboxing during the derivation build process.

### Steps To Reproduce
1. make a flake inside a repo with go code, and workaround #202592 explicitly:
```nix
{
  outputs = { nixpkgs, ... }: let
    system = "x86_64-darwin";
    pkgs = nixpkgs.legacyPackages.${system};
  in {
    checks.${system}.go = pkgs.runCommandLocal "go" {
      builtInputs = [pkgs.go];
    } "${pkgs.golangci-lint}/bin/golangci-lint run ${./.}";
  };
}
```
2. `nix flake check`
```
2022/11/23 21:13:36 failed to initialize build cache at /homeless-shelter/Library/Caches/golangci-lint: mkdir /homeless-shelter: read-only file system
```
3. set `GOLANGCI_LINT_CACHE=/tmp/`
4. `nix flake check`
```
ERRO Running error: context loading failed: failed to load packages: failed to load with go/packages: err: exit status 1: stderr: failed to initialize build cache at /homeless-shelter/Library/Caches/go-build: mkdir /homeless-shelter: read-only file system 
```
5. set `GOCACHE=/tmp/`
6. `nix flake check`
```
ERRO Running error: context loading failed: failed to load packages: failed to load with go/packages: err: exit status 1: stderr: package unsafe is not in GOROOT (/nix/store/1bj9310khph9p26wiad36m8x03qjxdpq-go-1.19.2/share/go/src/>
```

### Expected behavior
it just works:
```
```

### Additional context
it was required that I work around #202592 during this investigation.

I was able to resolve the two cache failures by explicitly setting `GOCACHE` and `GOLANGCI_LINT_CACHE` to a writable temporary directory, /tmp/ may not work well on linux.

I am unclear if setting these via wrapper is the right call, even in `stdenv`, since there are valid reasons why you would like these to point to a reused cache during normal operation. we could export some helper functions like `buildGoModule`, say `checkGolangCiLint`, since we will likely be heavy handed with the solution?

I think there may also be is an actual bug with the `GOROOT` stuff, but I could not find the root cause, so feel free to fork that off into a separate issue.

### Notify maintainers
golangci-lint: @anpryl @manveru @mic92
go: @c00w @kalbasit @mic92 @zowoq

### Metadata
```
 - system: `"x86_64-darwin"`
 - host os: `Darwin 21.6.0, macOS 10.16`
 - multi-user?: `yes`
 - sandbox: `no`
 - version: `nix-env (Nix) 2.11.0`
 - channels(user): `"darwin, home-manager"`
 - channels(root): `"nixpkgs"`
 - nixpkgs: `/nix/store/lxfwcg0km6p3rrr307i9k6pgnc4brmzb-source`
```

**Comment by Mic92** [comment]
This is my setup in sops-nix:

https://github.com/Mic92/sops-nix/blob/f72e050c3ef148b1131a0d2df55385c045e4166b/pkgs/lint.nix#L2

I only had to give it a $HOME to work.

**Comment by Mic92** [comment]
I think a `golangciLintHook` package could wrap this nicely. I would not add a build hook to the linter itself since this might draw in dependencies i.e. a go compiler.

**Comment by colin-arnott-xero** [comment]
yeah, as feared the rabbit hole goes one deeper, since `golangci-lint` basically needs access to the same vendor directory that `buildGoModule` uses. how did you get this working on sops-nix?

**Comment by xsteadfastx** [comment]
tried it today... nothing changed.

**Comment by Mic92** [comment]
> yeah, as feared the rabbit hole goes one deeper, since `golangci-lint` basically needs access to the same vendor directory that `buildGoModule` uses. how did you get this working on sops-nix?

I override the original package that was used for `buildGoModule`.
